src/filter.ts

```diff
@@ -19,208 +34,283 @@ function match(pattern: any, value: any) {
-    if (!device) return false;
-    return match(value, device.manufacturer);
-  },
-  device_model: async (hass, value, entity) => {
-    const ent = (await getEntities(hass)).find(
-      (e) => e.entity_id === entity.entity_id
-    );
-    if (!ent) return false;
-    const device = (await getDevices(hass)).find((d) => d.id === ent.device_id);
-    if (!device) return false;
-    return match(value, device.model);
-  },
-  area: async (hass, value, entity) => {
-    const ent = (await getEntities(hass)).find(
-      (e) => e.entity_id === entity.entity_id
-    );
-    if (!ent) return false;
-    let area = (await getAreas(hass)).find((a) => a.area_id === ent.area_id);
-    if (area) return match(value, area.name) || match(value, area.area_id);
-    const device = (await getDevices(hass)).find((d) => d.id === ent.device_id);
-    if (!device) return false;
-    area = (await getAreas(hass)).find((a) => a.area_id === device.area_id);
-    if (!area) return false;
-    return match(value, area.name) || match(value, area.area_id);
-  },
-  entity_category: async (hass, value, entity) => {
-    const ent = (await getEntities(hass)).find(
-      (e) => e.entity_id === entity.entity_id
-    );
-    if (!ent) return false;
-    return match(value, ent.entity_category);
+  device: async (hass, value) => {
+    const [entities, devices, matcher] = await Promise.all([
+      getEntities(hass),
+      getDevices(hass),
+      match(value),
+    ]);
+    return (entity) => {
+      const ent = entities.get(entity.entity_id);
+      if (!ent) return false;
+      const device = devices.get(ent.device_id);
+      if (!device) return false;
+      return matcher(device.name_by_user) || matcher(device.name);
+    };
+  },
+  device_manufacturer: async (hass, value) => {
+    const [entities, devices, matcher] = await Promise.all([
+      getEntities(hass),
+      getDevices(hass),
+      match(value),
+    ]);
+    return (entity) => {
+      const ent = entities.get(entity.entity_id);
+      if (!ent) return false;
+      const device = devices.get(ent.device_id);
+      if (!device) return false;
+      return matcher(device.manufacturer);
+    };
+  },
+  device_model: async (hass, value) => {
+    const [entities, devices, matcher] = await Promise.all([
+      getEntities(hass),
+      getDevices(hass),
+      match(value),
+    ]);
+    return (entity) => {
+      const ent = entities.get(entity.entity_id);
+      if (!ent) return false;
+      const device = devices.get(ent.device_id);
+      if (!device) return false;
+      return matcher(device.model);
+    };
+  },
+  area: async (hass, value) => {
+    const [entityAreas, areas, matcher] = await Promise.all([
+      getEntityAreas(hass),
+      getAreas(hass),
+      match(value),
+    ]);
+    return (entity) => {
+      const area = areas.get(entityAreas.get(entity.entity_id));
+      if (!area) return false;
+      return (
```

Comment:
Entities that have no assigned area can still belong to a device which has an area. Those should be matched too.